File: shapes.py
"""
shapes.py
Draws the six Qwirkle shapes (Circle, Star, Diamond, Square, Clover, X)
onto a pygame surface. Kept separate from gui.py so the art is easy to
tweak without wading through layout code.
"""

import math
import pygame

WHITE = (255, 255, 255)
BLACK = (20, 20, 20)


def _star_points(center, outer_r, inner_r, points=5, rotation=-90):
    cx, cy = center
    pts = []
    step = math.pi / points
    angle = math.radians(rotation)
    for i in range(points * 2):
        r = outer_r if i % 2 == 0 else inner_r
        pts.append((cx + r * math.cos(angle), cy + r * math.sin(angle)))
        angle += step
    return pts


def draw_shape(surface, shape, center, size, color, outline=BLACK, outline_w=2):
    """
    shape: one of Circle, Star, Diamond, Square, Clover, X
    center: (x, y) pixel center
    size: roughly the shape's radius/half-width in pixels
    """
    cx, cy = center

    if shape == "Circle":
        pygame.draw.circle(surface, color, (int(cx), int(cy)), int(size))
        pygame.draw.circle(surface, outline, (int(cx), int(cy)), int(size), outline_w)

    elif shape == "Square":
        rect = pygame.Rect(0, 0, size * 1.6, size * 1.6)
        rect.center = (cx, cy)
        pygame.draw.rect(surface, color, rect, border_radius=3)
        pygame.draw.rect(surface, outline, rect, outline_w, border_radius=3)

    elif shape == "Diamond":
        pts = [(cx, cy - size), (cx + size, cy), (cx, cy + size), (cx - size, cy)]
        pygame.draw.polygon(surface, color, pts)
        pygame.draw.polygon(surface, outline, pts, outline_w)

    elif shape == "Star":
        pts = _star_points((cx, cy), size, size * 0.42, points=5)
        pygame.draw.polygon(surface, color, pts)
        pygame.draw.polygon(surface, outline, pts, outline_w)

    elif shape == "X":
        w = max(3, int(size * 0.42))
        _thick_line(surface, color, (cx - size, cy - size), (cx + size, cy + size), w, outline, outline_w)
        _thick_line(surface, color, (cx - size, cy + size), (cx + size, cy - size), w, outline, outline_w)

    elif shape == "Clover":
        # A 3-leaf clover: three overlapping circular "leaves" arranged in a
        # triangle around the center, plus a small stem.
        leaf_r = size * 0.55
        offset = size * 0.55
        angles = [-90, 30, 150]  # degrees, evenly spaced (3 leaves)
        for a in angles:
            rad = math.radians(a)
            lx = cx + offset * math.cos(rad)
            ly = cy + offset * math.sin(rad)
            pygame.draw.circle(surface, color, (int(lx), int(ly)), int(leaf_r))
        for a in angles:
            rad = math.radians(a)
            lx = cx + offset * math.cos(rad)
            ly = cy + offset * math.sin(rad)
            pygame.draw.circle(surface, outline, (int(lx), int(ly)), int(leaf_r), outline_w)

    else:
        raise ValueError(f"Unknown shape: {shape}")


def _thick_line(surface, color, p1, p2, width, outline, outline_w):
    x1, y1 = p1
    x2, y2 = p2
    dx, dy = x2 - x1, y2 - y1
    length = math.hypot(dx, dy) or 1
    nx, ny = -dy / length * (width / 2), dx / length * (width / 2)
    pts = [
        (x1 + nx, y1 + ny), (x2 + nx, y2 + ny),
        (x2 - nx, y2 - ny), (x1 - nx, y1 - ny),
    ]
    pygame.draw.polygon(surface, color, pts)
    pygame.draw.polygon(surface, outline, pts, max(1, outline_w - 1))

File: js/shapes.js
// shapes.js
// Draws the six Qwirkle shapes (Circle, Star, Diamond, Square, Clover,
// Triangle) onto a canvas 2D context. Kept separate from gui.js so the
// art is easy to tweak without wading through layout code.
// Port of shapes.py (pygame.draw -> Canvas 2D).

const OUTLINE = "rgb(20, 20, 20)";

function starPoints(cx, cy, outerR, innerR, points = 5, rotationDeg = -90) {
  const pts = [];
  const step = Math.PI / points;
  let angle = (rotationDeg * Math.PI) / 180;
  for (let i = 0; i < points * 2; i++) {
    const r = i % 2 === 0 ? outerR : innerR;
    pts.push([cx + r * Math.cos(angle), cy + r * Math.sin(angle)]);
    angle += step;
  }
  return pts;
}

function fillStrokePolygon(ctx, pts, color, outline, outlineW) {
  ctx.beginPath();
  ctx.moveTo(pts[0][0], pts[0][1]);
  for (let i = 1; i < pts.length; i++) ctx.lineTo(pts[i][0], pts[i][1]);
  ctx.closePath();
  ctx.fillStyle = color;
  ctx.fill();
  ctx.strokeStyle = outline;
  ctx.lineWidth = outlineW;
  ctx.stroke();
}

// shape: one of Circle, Star, Diamond, Square, Clover, Triangle
// cx, cy: pixel center
// size: roughly the shape's radius/half-width in pixels
export function drawShape(ctx, shape, cx, cy, size, color, outline = OUTLINE, outlineW = 2) {
  switch (shape) {
    case "Circle": {
      ctx.beginPath();
      ctx.arc(cx, cy, size, 0, Math.PI * 2);
      ctx.fillStyle = color;
      ctx.fill();
      ctx.strokeStyle = outline;
      ctx.lineWidth = outlineW;
      ctx.stroke();
      break;
    }

    case "Square": {
      const s = size * 1.6;
      const x = cx - s / 2, y = cy - s / 2;
      const r = 3;
      roundRectPath(ctx, x, y, s, s, r);
      ctx.fillStyle = color;
      ctx.fill();
      ctx.strokeStyle = outline;
      ctx.lineWidth = outlineW;
      ctx.stroke();
      break;
    }

    case "Diamond": {
      const pts = [
        [cx, cy - size],
        [cx + size, cy],
        [cx, cy + size],
        [cx - size, cy],
      ];
      fillStrokePolygon(ctx, pts, color, outline, outlineW);
      break;
    }

    case "Star": {
      const pts = starPoints(cx, cy, size, size * 0.42, 5);
      fillStrokePolygon(ctx, pts, color, outline, outlineW);
      break;
    }

    case "Triangle": {
      const h = size * 1.3;
      const pts = [
        [cx, cy - h * 0.62],
        [cx - size, cy + h * 0.38],
        [cx + size, cy + h * 0.38],
      ];
      fillStrokePolygon(ctx, pts, color, outline, outlineW);
      break;
    }

    case "Clover": {
      // A 3-leaf clover: three overlapping circular "leaves" arranged in
      // a triangle around the center.
      const leafR = size * 0.55;
      const offset = size * 0.55;
      const angles = [-90, 30, 150]; // degrees, evenly spaced (3 leaves)
      const leaves = angles.map((a) => {
        const rad = (a * Math.PI) / 180;
        return [cx + offset * Math.cos(rad), cy + offset * Math.sin(rad)];
      });
      for (const [lx, ly] of leaves) {
        ctx.beginPath();
        ctx.arc(lx, ly, leafR, 0, Math.PI * 2);
        ctx.fillStyle = color;
        ctx.fill();
      }
      for (const [lx, ly] of leaves) {
        ctx.beginPath();
        ctx.arc(lx, ly, leafR, 0, Math.PI * 2);
        ctx.strokeStyle = outline;
        ctx.lineWidth = outlineW;
        ctx.stroke();
      }
      break;
    }

    default:
      throw new Error(`Unknown shape: ${shape}`);
  }
}

// Manual rounded-rect path (works even on the rare browser without
// CanvasRenderingContext2D.roundRect).
export function roundRectPath(ctx, x, y, w, h, r) {
  const rr = Math.min(r, w / 2, h / 2);
  ctx.beginPath();
  ctx.moveTo(x + rr, y);
  ctx.lineTo(x + w - rr, y);
  ctx.arcTo(x + w, y, x + w, y + rr, rr);
  ctx.lineTo(x + w, y + h - rr);
  ctx.arcTo(x + w, y + h, x + w - rr, y + h, rr);
  ctx.lineTo(x + rr, y + h);
  ctx.arcTo(x, y + h, x, y + h - rr, rr);
  ctx.lineTo(x, y + rr);
  ctx.arcTo(x, y, x + rr, y, rr);
  ctx.closePath();
}
